client/beefy/src/worker.rs

```diff
@@ -447,19 +459,29 @@ where
 	/// 3. Send best block hash and `finality_proof` to RPC worker.
 	///
 	/// Expects `finality proof` to be valid.
-	fn finalize(&mut self, finality_proof: BeefyVersionedFinalityProof<B>) {
+	fn finalize(&mut self, finality_proof: BeefyVersionedFinalityProof<B>) -> Result<(), Error> {
+		let block_num = match finality_proof {
+			VersionedFinalityProof::V1(ref sc) => sc.commitment.block_number,
+		};
+
+		// Conclude voting round for this block.
+		self.voting_oracle.rounds_mut().ok_or(Error::UninitSession)?.conclude(block_num);
 		// Prune any now "finalized" sessions from queue.
 		self.voting_oracle.try_prune();
-		let signed_commitment = match finality_proof {
-			VersionedFinalityProof::V1(ref sc) => sc,
-		};
-		let block_num = signed_commitment.commitment.block_number;
+
 		if Some(block_num) > self.best_beefy_block {
 			// Set new best BEEFY block number.
 			self.best_beefy_block = Some(block_num);
 			metric_set!(self, beefy_best_block, block_num);
 
-			self.client.hash(block_num).ok().flatten().map(|hash| {
+			if let Err(e) = self.backend.append_justification(
```

Comment:
Isn't this operation success relevant? 
In other words, why is the success of `append_justification` considered "optional"?

Reply:
It's not that success here is "optional" more that failure is not fatal. The voter can still make progress so it seems extreme to kill it if justification(s) can't be appended to backend.
So we just log an error.

We should change `debug!()` to `error!()` here, though, good catch!